src/vmm/src/devices/virtio/net/device.rs

```diff
@@ -644,6 +642,18 @@ impl Net {
         }
     }
 
+    fn build_tap_supported_features(desired_flags: &[(u32, u32)], supported_features: u64) -> u32 {
+        let mut answer: u32 = 0;
+
+        for (tap_flag, net_flag) in desired_flags.iter() {
+            if (supported_features & 1 << net_flag) != 0 {
+                answer |= tap_flag;
+            }
+        }
+
+        return answer;
+    }
+
```

Comment:
personally, I would keep this much simpler. Something like:

```Rust
fn build_tap_supported_features(&self, supported_features: u64) -> u32 {
    let mut tap_features: u32 = 0;

    if self.has_features(VIRTIO_NET_F_GUEST_TSO4) {
        tap_features |= gen::TUN_F_TSO4;
    }
    if self.has_feature(....) {

    }
}
```

and then later during activate, you don't have to build the array, you just:

```Rust
        let supported_flags =
            Net::build_tap_supported_features(self.acked_features());
```

Reply:
Tbh, I'm not sure it's better in term of testing. In this case we won't be able to write unit tests with extra / less flags we want to set up.

Reply:
I think it would. The only purpose the `desired_flags` parameter serves is to "build" the correlation between VirtIO and TAP flags. You can do exactly the same test you are doing in this PR, without the need to pass around the `desired_flags` array. This is simpler, since you only need to encode knowledge for the TAP features in this function here. For example:

```rust
    #[test]
    fn test_build_tap_supported_features_ack_les_than_want() {
        let tap_features: u32 = gen::TUN_F_CSUM | gen::TUN_F_UFO;
        let ack_features: u32 = 1 << VIRTIO_NET_F_CSUM | 1 << VIRTIO_NET_F_GUEST_UFO;

        let supported_flags =
            Net::build_tap_supported_features(ack_features as u64);

        assert_eq!(supported_flags, tap_features);
    }
```

Reply:
The issue with `if self.has_features(VIRTIO_NET_F_GUEST_TSO4) {` approach is that the method does not have `self` parameter. I think approach with map/array as a constant will be ok. Also I don't think we really need to test what happens if this `desired_flags` has incorrect/incomplete  entries.

Reply:
> The issue with if self.has_features(VIRTIO_NET_F_GUEST_TSO4) { approach is that the method does not have self parameter

Let's just add it then :)

Reply:
Then in the tests we will need to create `net` devices. And all of this just to not write `if (supported_features & 1 << net_flag) != 0`.

Reply:
alternatively, it doesn't have to make use of the `has_feature` it can just be:

```rust
if (flags & (1 << VIRTIO_NET_F_CSUM)) {
    tap_flags |= gen::TUN_F_CSUM;
}
```

Reply:
Just to clarify, I'm just making this suggestion because I'd like to avoid the for loop and keeping track of this constant array. I think having an array is an overkill and looping might hurt performance

Reply:
It seems we cycled back to

Testing each flag in separate `if`:
```rust
if (flags & (1 << VIRTIO_NET_F_CSUM)) {
    tap_flags |= gen::TUN_F_CSUM;
}
...
```
vs

Testing all in one loop:
```rust
const FEATURE_MAP = [
  (VIRTIO_NET_F_CSUM, TUN_F_CSUM),
  ...
  ]

...
for (tap_flag, net_flag) in desired_flags.iter() {
        if (supported_features & 1 << net_flag) != 0 {
            answer |= tap_flag;
        }
}
```

I am fine with any solution here.

Reply:
Yup, I'd rather go for the first approach.

Reply:
Oh, well. Here are my 50 cents:

1. If we don't care about testing it at all we can inline `build_tap_supported_features()` into `activate()` function itself. With or without the array.
2. In general I'd suggest to have a  `convert_tap_to_vitio_future(tap_feat: u32) -> u32` function, which will own converting logic. Cos as it was fairly mentioned - we use the same kind of arrays in different places. I'm not sure where should it be though (in `if_tun.rs` or `virtio_net.rs`.
3. It's nit, but I'd like to see function `convert_vitio_feat_to_bitflag()`, which makes `1 << flag`, cos it's a little bit mess to have it everywhere around the code. Moreover, tap features' consts are bitflags and virtio flags are not. (which can be fixed instead of having the function). But I understand it may affect a lot of code and may be undesirable. 

In general I would vote for array implementation, cos:
1. For now we have few features. But if it was 30? 50? after what number you would convert it into an array?
2. It would be nice to have all tap feature we'd like to setup in one place for readability (maybe as a static field of the class or as const array in the top of the file). To not seek it in the code, when we need to change flags we try to set up.

Reply:
I would prefer to keep complexity and amount of indirection to minimum. The whole conversion is a simple bit checking and we don't need to create a huge machinery to do that.
The amount of features it not really a concern here. It does not really matter if it will be 50 lines of array definition with all mappings or 50 lines of bit checks. Both will be +- same size and complexity. In the end it is to to the tests and person doing review to check that all mapping as correct.

I think smth simple as this should be enough. No additional arrays, statics... just one pure function.
```rust
fn virtio_to_tap_features(virtio_features: u64) -> u32 {
    let check_set = |virtio_features, virtio_feature, &mut tap_features, tap_feature| {
        if (virtio_features & (1 << virtio_feature)) {
            tap_features |= tap_feature;
        }
    };

    let mut tap_features = 0;

    check_set(virtio_features, VIRTIO_NET_F_CSUM, &mut tap_features, TUN_F_CSUM);
    check_set(virtio_features, VIRTIO_NET_F_GUEST_UFO, &mut tap_features, TUN_F_UFO);
    check_set(virtio_features, VIRTIO_NET_F_GUEST_TSO4, &mut tap_features, TUN_F_TSO4);
    check_set(virtio_features, VIRTIO_NET_F_GUEST_TSO6, &mut tap_features, TUN_F_TSO6);

    tap_features
}
```

Reply:
as you wish